### DISCUSSION TITLE
why does running `rg` with `sudo` on `/` cause my Linux system to hang?

### DISCUSSION THREAD

**Discussion by mgiv** [BurntSushi/ripgrep#2897] [Q&A] [answered]
### Please tick this box to confirm you have reviewed the above.

- [X] I have a different issue.

### What version of ripgrep are you using?

ripgrep 14.1.1

### How did you install ripgrep?

pacman

### What operating system are you using ripgrep on?

Arch Linux

### Describe your bug.

Without using sudo, searching in / will come up with errors such as:
```
rg: /proc/1411/task/1411/net/nf_conntrack_expect: Permission denied (os error 13)
```
However, there seems to be others, such as Input/output error, Invalid argument, etc.

To avoid this I added sudo, which works on things like find. However with rg it completely hangs the system and gets stuck on something in /proc

This is only recoverable by force restarting the computer

### What are the steps to reproduce the behavior?

`sudo rg foobar /` from anywhere in the filesystem

### What is the actual behavior?

[Logs](https://gist.github.com/bluescorpion35/66d5ad0285ad833affe768e341585292)

Note: the actual output is longer, the terminal only showed a certain amount of it. Also I killed rg after a few seconds, in case it crashed the system again

### What is the expected behavior?

Either not shown the errors, or at least not crash the entire system

I think rg should entirely ignore /proc unless an option is specified

**Comment by mgiv** [comment]
Error logs from journalctl
```
Sep 18 20:41:10 archlinux kernel: [Hardware Error]: System Fatal error.
Sep 18 20:41:10 archlinux kernel: [Hardware Error]: CPU:2 (19:21:0) MC5_STATUS[-|UE|MiscV|AddrV|PCC|TCC|SyndV|-|-|-]: 0xbea0000001000108
Sep 18 20:41:10 archlinux kernel: [Hardware Error]: Error Addr: 0x00ffffffc1315e4e
Sep 18 20:41:10 archlinux kernel: [Hardware Error]: IPID: 0x000500b000000000, Syndrome: 0x000000004d000000
Sep 18 20:41:10 archlinux kernel: [Hardware Error]: Execution Unit Ext. Error Code: 0
Sep 18 20:41:10 archlinux kernel: [Hardware Error]: cache level: RESV, tx: GEN, mem-tx: GEN
```

**Comment by BurntSushi** [comment, marked as answer]
This is generally expected behavior. Linux exposes special files in `/proc` where even the mere act of reading them can cause side effects, including system hangs or reboots.

The suggestion to ignore `/proc` by default is interesting, but I don't think it's viable to do that at this stage in ripgrep's life. ripgrep's smart filtering by default is already surprising in some respects, and adding more special cases isn't really something I'm willing to do. You can work around this by adding a `/.rgignore` to your system excluding `/proc`. You might also want to exclude `/dev` and `/sys`.

**Reply by mgiv** [reply]
Thanks

**Reply by adsick** [reply]
is `/.rignore` a thing? asking because I can't google a reference to it.
Also I'd ripgrep to not search in `/proc` and other special dirs as well, it makes sense only in very specific scenarios (which I actually worked with back in the day), but most people are more like "Well I want to find that file, but I don't know where even search... oh shit, there are a ton of permission denied messages! hm, I may try to run it with sudo(1)... bang - system crash." - not a very pleasant experience.
(1) I think modern linux sort of teaches users to abuse sudo every time

**Reply by adsick** [reply]
"/proc" search should be opt-in, not opt-out imho.

also for some reason I still was getting permission denied messages when using these options:
```
--glob=!/proc/**
--glob=!/sys/**
--glob=!/dev/**
--glob=!/run/**
--no-follow
```

**Reply by BurntSushi** [reply]
Yes, you can use `/.rgignore`.
